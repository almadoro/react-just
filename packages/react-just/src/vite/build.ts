import path from "node:path";
import type { Plugin } from "vite";
import { getRegisterModuleIdFromPath } from "./utils/client";

type BuildOptions = { entry: string };

export default function build(options: BuildOptions): Plugin {
  const clientEntries: string[] = [];

  return {
    name: "react-just:build",
    apply: "build",
    config() {
      return {
        appType: "custom",
        builder: {
          sharedConfigBuild: true,
          async buildApp(builder) {
            await builder.build(builder.environments.server);
            // It is required that the server build is executed first to catch all
            // the client entries.
            await builder.build(builder.environments.client);
          },
        },
        environments: {
          server: {
            consumer: "server",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              outDir: "dist/server",
              rollupOptions: {
                input: { index: path.resolve(options.entry) },
                output: { entryFileNames: "index.js" },
              },
            },
          },
          client: {
            consumer: "client",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              minify: false,
              outDir: "dist/client",
              rollupOptions: {
                input: { index: ENTRY_MODULE_ID },
                output: { entryFileNames: "index.js" },
              },
            },
          },
        },
      };
    },
    moduleParsed(moduleInfo) {
      if (this.environment.name !== "server") return;

      const isClientFile = moduleInfo.meta.reactUseClient?.transformed;
      if (isClientFile) clientEntries.push(moduleInfo.id);
    },
    resolveId(id) {
      if (id === ENTRY_MODULE_ID) return RESOLVED_ENTRY_MODULE_ID;
    },
    load(id) {
      if (id === RESOLVED_ENTRY_MODULE_ID)
        return getClientEntryCode(clientEntries);
    },
  };
}

export const ENTRY_MODULE_ID = "/virtual:client-entry";

const RESOLVED_ENTRY_MODULE_ID = "\0" + ENTRY_MODULE_ID;

function getClientEntryCode(entries: string[]) {
  let code = `import { hydrateFromWindowFlight, registerModule } from "react-just/client";`;

  for (let i = 0; i < entries.length; i++) {
    const entryPath = entries[i];
    const moduleId = getRegisterModuleIdFromPath(entryPath);
    code += `import * as entry_${i} from "${entryPath}";`;
    code += `registerModule("${moduleId}", entry_${i});`;
  }

  code += "hydrateFromWindowFlight();";
  return code;
}
