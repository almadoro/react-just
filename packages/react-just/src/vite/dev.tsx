import path from "node:path";
import type {
  Connect,
  PluginOption,
  RunnableDevEnvironment,
  ViteDevServer,
} from "vite";
import {
  renderToFlightPipeableStream,
  renderToHtmlPipeableStream,
} from "../../types/server.node";

type DevOptions = { app: string; flightMimeType: string };

export default function dev(options: DevOptions): PluginOption {
  return {
    name: "react-just:dev",
    apply: "serve",
    configureServer(server) {
      return () =>
        server.middlewares.use(middleware(server, options.flightMimeType));
    },
    resolveId(id) {
      if (id === CLIENT_ENTRY_MODULE_ID) return RESOLVED_CLIENT_ENTRY_MODULE_ID;
      if (id === SERVER_ENTRY_MODULE_ID) return RESOLVED_SERVER_ENTRY_MODULE_ID;
    },
    load(id) {
      if (id === RESOLVED_CLIENT_ENTRY_MODULE_ID) return CLIENT_ENTRY;
      if (id === RESOLVED_SERVER_ENTRY_MODULE_ID)
        return getServerEntry(
          path.resolve(this.environment.config.root, options.app),
        );
    },
  };
}

const CLIENT_ENTRY_MODULE_ID = "/virtual:client-entry";

const RESOLVED_CLIENT_ENTRY_MODULE_ID = "\0" + CLIENT_ENTRY_MODULE_ID;

const HMR_CODE =
  // Taken from: https://github.com/vitejs/vite-plugin-react/blob/main/packages/common/refresh-utils.ts
  `import { injectIntoGlobalHook } from "/@react-refresh";` +
  `injectIntoGlobalHook(window);` +
  `window.$RefreshReg$ = () => {};` +
  `window.$RefreshSig$ = () => (type) => type;`;

const CLIENT_ENTRY =
  HMR_CODE +
  `import { hydrateFromWindowFlight } from "react-just/client";` +
  `hydrateFromWindowFlight();`;

const SERVER_ENTRY_MODULE_ID = "/virtual:server-entry";

const RESOLVED_SERVER_ENTRY_MODULE_ID = "\0" + SERVER_ENTRY_MODULE_ID;

type RenderToHtmlPipeableStream = typeof renderToHtmlPipeableStream;
type RenderToFlightPipeableStream = typeof renderToFlightPipeableStream;

function middleware(
  server: ViteDevServer,
  flightMimeType: string,
): Connect.NextHandleFunction {
  return async (req, res) => {
    const ssr = server.environments.ssr as RunnableDevEnvironment;

    const serverModule = await ssr.runner.import(SERVER_ENTRY_MODULE_ID);

    const { renderToFlightPipeableStream, renderToHtmlPipeableStream, App } =
      serverModule as {
        renderToFlightPipeableStream: RenderToFlightPipeableStream;
        renderToHtmlPipeableStream: RenderToHtmlPipeableStream;
        App: React.ComponentType;
      };

    const AppRoot = () => (
      <>
        <script async type="module" src={CLIENT_ENTRY_MODULE_ID} />
        <App />
      </>
    );

    if (req.headers.accept?.includes(flightMimeType)) {
      res.statusCode = 200;
      res.setHeader("content-type", flightMimeType);
      const flightStream = renderToFlightPipeableStream(<AppRoot />);
      flightStream.pipe(res);
      return;
    }

    res.statusCode = 200;
    res.setHeader("content-type", "text/html");
    const htmlStream = renderToHtmlPipeableStream(<AppRoot />);
    htmlStream.pipe(res);
  };
}

function getServerEntry(appPath: string) {
  return (
    `import { renderToFlightPipeableStream, renderToHtmlPipeableStream } from "react-just/server.node";` +
    `import App from "${path.resolve(appPath)}";` +
    `export { renderToFlightPipeableStream, renderToHtmlPipeableStream, App } `
  );
}
