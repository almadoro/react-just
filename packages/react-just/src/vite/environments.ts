import {
  createServerHotChannel,
  createServerModuleRunner,
  DevEnvironment,
  Plugin,
  ResolvedConfig,
} from "vite";
import {
  ModuleEvaluator,
  ModuleRunner,
  ModuleRunnerContext,
} from "vite/module-runner";

export default function environments(): Plugin {
  return {
    name: "react-just:environments",
    config() {
      return {
        // noDiscovery: false is required to be able to trigger
        // reoptimizations and analyze specified entries.
        environments: {
          [ENVIRONMENTS.CLIENT]: {
            consumer: "client",
            optimizeDeps: {
              noDiscovery: false,
            },
          },
          [ENVIRONMENTS.FIZZ_NODE]: {
            consumer: "server",
            resolve: {
              noExternal: true,
            },
            optimizeDeps: {
              noDiscovery: false,
            },
          },
          [ENVIRONMENTS.FLIGHT_NODE]: {
            consumer: "server",
            resolve: {
              conditions: ["react-server"],
              noExternal: true,
            },
            optimizeDeps: {
              esbuildOptions: { conditions: ["react-server"] },
              noDiscovery: false,
            },
          },
          // On scan environments, no optimization is required since the code
          // won't be executed.
          [ENVIRONMENTS.SCAN_USE_CLIENT_MODULES]: {
            consumer: "server",
            resolve: {
              conditions: ["react-server"],
              // Since "use client" directive might appear in dependencies, we
              // can't treat them as external.
              noExternal: true,
            },
            optimizeDeps: {
              noDiscovery: true,
            },
            dev: {
              createEnvironment(name, config) {
                return new ScanEnvironment(name, config);
              },
            },
          },
          [ENVIRONMENTS.SCAN_USE_SERVER_MODULES]: {
            consumer: "server",
            resolve: {
              // Since the "use server" directive won't appear in dependencies,
              // we can treat them as external.
              external: true,
            },
            optimizeDeps: {
              noDiscovery: true,
            },
            dev: {
              createEnvironment(name, config) {
                return new ScanEnvironment(name, config);
              },
            },
          },
        },
      };
    },
  };
}

export const ENVIRONMENTS = {
  CLIENT: "client",
  FIZZ_NODE: "fizz_node",
  FLIGHT_NODE: "flight_node",
  SCAN_USE_CLIENT_MODULES: "scan_use_client_modules",
  SCAN_USE_SERVER_MODULES: "scan_use_server_modules",
};

/**
 * @returns Whether the environment is a client or fizz environment.
 */
export function isClientLikeEnvironment(environment: string) {
  return environment === ENVIRONMENTS.CLIENT || isFizzEnvironment(environment);
}

export function isFlightEnvironment(environment: string) {
  return environment === ENVIRONMENTS.FLIGHT_NODE;
}

export function isScanUseClientModulesEnvironment(environment: string) {
  return environment === ENVIRONMENTS.SCAN_USE_CLIENT_MODULES;
}

export function isScanUseServerModulesEnvironment(environment: string) {
  return environment === ENVIRONMENTS.SCAN_USE_SERVER_MODULES;
}

function isFizzEnvironment(environment: string) {
  return environment === ENVIRONMENTS.FIZZ_NODE;
}

// ScanEnvironment triggers `transform` hook in the same way as the
// RunnableDevEnvironment but it doesn't execute the code. To achieve that, we
// use a custom ModuleEvaluator that only triggers vite imports calls.
export class ScanEnvironment extends DevEnvironment {
  private runner: ModuleRunner;

  constructor(name: string, config: ResolvedConfig) {
    super(name, config, { hot: false, transport: createServerHotChannel() });

    this.runner = createServerModuleRunner(this, {
      hmr: false,
      evaluator: new ImportsEvaluator(),
    });
  }

  public async scan(id: string) {
    await this.runner.import(id);
  }
}

class ImportsEvaluator implements ModuleEvaluator {
  public async runExternalModule() {}

  public async runInlinedModule(
    { __vite_ssr_import__ }: ModuleRunnerContext,
    code: string,
  ) {
    const ids = new Set<string>();

    let match: RegExpExecArray | null;

    const importRegex = /__vite_ssr_import__\(\s*["'`]([^"'`]+)["'`][^)]*\)/g;
    while (true) {
      match = importRegex.exec(code);
      if (!match) break;
      ids.add(match[1]);
    }

    const dynamicImportRegex =
      /__vite_ssr_dynamic_import__\(\s*["'`]([^"'`]+)["'`][^)]*\)/g;
    while (true) {
      match = dynamicImportRegex.exec(code);
      if (!match) break;
      ids.add(match[1]);
    }

    await Promise.all(Array.from(ids).map((id) => __vite_ssr_import__(id)));
  }
}
