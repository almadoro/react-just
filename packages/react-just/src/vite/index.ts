import react from "@vitejs/plugin-react";
import type { PluginOption } from "vite";
import build from "./build";
import dev from "./dev";
import { ReactJustOptions } from "./types";
import useClient from "./use-client";
import useServer from "./use-server";

const FLIGHT_MIME_TYPE = "text/x-component";

export default function reactJust(options: ReactJustOptions): PluginOption {
  return [
    build({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    dev({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    react(),
    useClient(),
    useServer(),
  ];
}
