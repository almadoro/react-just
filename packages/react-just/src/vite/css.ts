import { DevEnvironment, Plugin } from "vite";
import { isFlightEnvironment } from "./environments";
import { isCssModule } from "./utils";

export default function css(): Plugin {
  const flightEnvironments: DevEnvironment[] = [];

  return {
    name: "react-just:css",
    configureServer(server) {
      for (const environment of Object.values(server.environments)) {
        if (isFlightEnvironment(environment.name))
          flightEnvironments.push(environment);
      }
    },
    resolveId(id) {
      if (id === CSS_MODULES) return RESOLVED_CSS_MODULES;
    },
    load(id) {
      if (id === RESOLVED_CSS_MODULES)
        return getCssModulesCode(flightEnvironments);
    },
  };
}

export const CSS_MODULES = "/virtual:react-just/css-modules";
export const RESOLVED_CSS_MODULES = "\0" + CSS_MODULES;

function getCssModulesCode(flightEnvironments: DevEnvironment[]) {
  const ids = new Set<string>();

  for (const environment of flightEnvironments) {
    const { idToModuleMap } = environment.moduleGraph;

    for (const [id, module] of idToModuleMap.entries()) {
      if (
        isCssModule(id) &&
        // On development, map may contain unused modules. Ignore them.
        module.importers.size !== 0
      )
        ids.add(id);
    }
  }

  return Array.from(ids)
    .map((id) => `import "${id}";`)
    .join("\n");
}
