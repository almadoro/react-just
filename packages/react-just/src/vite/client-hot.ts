import { DevEnvironment, Plugin } from "vite";
import { CSS_MODULES } from "./css";
import { ENVIRONMENTS, isFlightEnvironment } from "./environments";
import { CLIENT_MODULES } from "./use-client";

export default function clientHot(): Plugin {
  return {
    name: "react-just:client-hot",
    configureServer(server) {
      watchClientReactChange(server.environments[ENVIRONMENTS.CLIENT]);
    },
    hotUpdate(ctx) {
      if (isFlightEnvironment(this.environment.name))
        onFlightHotUpdate(ctx.server.environments[ENVIRONMENTS.CLIENT]);
    },
    resolveId(id) {
      switch (id) {
        case CLIENT_HOT_MODULES:
          return RESOLVED_CLIENT_HOT_MODULES;
        case BROWSER_HASH_CHECK:
          return RESOLVED_BROWSER_HASH_CHECK;
        case HMR_PREAMBLE:
          return RESOLVED_HMR_PREAMBLE;
        case SERVER_HMR:
          return RESOLVED_SERVER_HMR;
      }
    },
    load(id) {
      switch (id) {
        case RESOLVED_CLIENT_HOT_MODULES:
          return getClientHotModulesCode();
        case RESOLVED_BROWSER_HASH_CHECK:
          if (!(this.environment instanceof DevEnvironment)) return "";
          const browserHash =
            this.environment.depsOptimizer?.metadata.browserHash;
          if (!browserHash)
            throw new Error("Expected browser hash to be available");
          return getBrowserHashCheckCode(browserHash);
        case RESOLVED_HMR_PREAMBLE:
          return getHmrPreambleCode();
        case RESOLVED_SERVER_HMR:
          return getServerHmrCode();
      }
    },
  };
}

const BROWSER_HASH_CHECK_HOT_EVENT = "react-just:browser-hash-check";

function watchClientReactChange(clientEnv: DevEnvironment) {
  let currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

  // When a reoptimization occurs, the React version of loaded modules on
  // client will not match future loaded ones. A full reload is required.
  setInterval(() => {
    let prevBrowserHash = currentBrowserHash;
    currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

    const needsReload = prevBrowserHash !== currentBrowserHash;
    if (needsReload) clientEnv.hot.send({ type: "full-reload" });
  }, 0);

  // A reoptimization may occur before the Vite connection is established;
  // thus, any triggered full-reload events won't be received. A manual check
  // is required after the connection is established.
  clientEnv.hot.on("connection", () => {
    // The event is sent to every connected client. This doesn't matter.
    // This event should have no effects when the browser's modules are up to
    // date.
    clientEnv.hot.send({
      type: "custom",
      event: BROWSER_HASH_CHECK_HOT_EVENT,
      data: { browserHash: currentBrowserHash },
    });
  });
}

const TREE_RELOAD_HOT_EVENT = "react-just:tree-reload";

function onFlightHotUpdate(clientEnv: DevEnvironment) {
  // Send an event to trigger the client modules and react tree reload.
  clientEnv.hot.send({
    type: "custom",
    event: TREE_RELOAD_HOT_EVENT,
    data: { eventId: crypto.randomUUID() },
  });
}

export const CLIENT_HOT_MODULES = "/virtual:react-just/client-hot-modules";
const RESOLVED_CLIENT_HOT_MODULES = "\0" + CLIENT_HOT_MODULES;

function getClientHotModulesCode() {
  return (
    `import "${BROWSER_HASH_CHECK}";` +
    `import "${HMR_PREAMBLE}";` +
    `import "${SERVER_HMR}";`
  );
}

const BROWSER_HASH_CHECK = "/virtual:react-just/browser-hash-check";
const RESOLVED_BROWSER_HASH_CHECK = "\0" + BROWSER_HASH_CHECK;

function getBrowserHashCheckCode(currentBrowserHash: string) {
  return (
    `if (import.meta.hot) {` +
    `  import.meta.hot.on("${BROWSER_HASH_CHECK_HOT_EVENT}", ({ browserHash }) => {` +
    `    if (browserHash !== "${currentBrowserHash}")` +
    `      window.location.reload();` +
    `  });` +
    `}`
  );
}

const HMR_PREAMBLE = "/virtual:react-just/hmr-preamble";
const RESOLVED_HMR_PREAMBLE = "\0" + HMR_PREAMBLE;

function getHmrPreambleCode() {
  // Taken from: https://github.com/vitejs/vite-plugin-react/blob/main/packages/common/refresh-utils.ts
  return (
    `import { injectIntoGlobalHook } from "/@react-refresh";` +
    `if (import.meta.hot) {` +
    `  injectIntoGlobalHook(window);` +
    `  window.$RefreshReg$ = () => {};` +
    `  window.$RefreshSig$ = () => (type) => type;` +
    `}`
  );
}

const SERVER_HMR = "/virtual:react-just/server-hmr";
const RESOLVED_SERVER_HMR = "\0" + SERVER_HMR;

function getServerHmrCode() {
  return (
    `import { createFromRscFetch, render, RSC_MIME_TYPE } from "react-just/client";` +
    `if (import.meta.hot) {` +
    `  let lastEventId = null;` +
    `  import.meta.hot.on("${TREE_RELOAD_HOT_EVENT}", ({ eventId }) => {` +
    `    lastEventId = eventId;` +
    `    const headers = { accept: RSC_MIME_TYPE };` +
    // Don't use exactly the same URL as the one we're trying to load to avoid
    // sharing cache between the RSC and the HTML.
    `    const url = new URL(window.location.href);` +
    `    url.searchParams.set("__rsc__", "1");` +
    `    createFromRscFetch(fetch(url, { headers })).then(async (tree) => {` +
    // Add a timestamp to trigger modules reload on the browser.
    `      await import(/* @vite-ignore */ "${CLIENT_MODULES}?t=" + Date.now());` +
    `      await import(/* @vite-ignore */ "${CSS_MODULES}?t=" + Date.now());` +
    // Avoid race conditions between multiple reloads. Render only the latest one.
    `      if (lastEventId === eventId) render(tree);` +
    `    });` +
    `  });` +
    `}`
  );
}
