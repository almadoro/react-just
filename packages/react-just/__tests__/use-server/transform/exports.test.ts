import baseTransform, { TransformOptions } from "@/vite/use-server/transform";
import Generator from "@/vite/use-server/transform/generator";
import { generate } from "astring";
import { builders } from "estree-toolkit";
import fs from "node:fs/promises";
import path from "node:path";
import { parseAst } from "vite";
import { describe, expect, test } from "vitest";

// flight-like environment
const noTreeshakeEnv: TransformOptions = {
  generator: new Generator({
    getRegisterArguments: ({ exportName, implementationIdentifier }) => [
      builders.identifier(implementationIdentifier),
      builders.literal("action:" + exportName),
      builders.literal(exportName),
    ],
    registerServerReferenceSource: "react-just/no-treeshake",
  }),
  treeshakeImplementation: false,
};

// client-like environment
const treeshakeEnv: TransformOptions = {
  generator: new Generator({
    getRegisterArguments: ({ exportName }) => [
      builders.literal("action:" + exportName),
    ],
    registerServerReferenceSource: "react-just/treeshake",
  }),
  treeshakeImplementation: true,
};

describe(
  "transforms exports on a no-treeshake environment",
  exportsTests(noTreeshakeEnv),
);

describe(
  "transforms exports on a treeshake environment",
  exportsTests(treeshakeEnv),
);

function exportsTests(options: TransformOptions) {
  return () => {
    describe("export default declarations", () => {
      test(
        "class declarations",
        transformTest("export-default-declaration/class-declaration.js"),
      );

      test(
        "function declarations",
        transformTest("export-default-declaration/function-declaration.js"),
      );

      test(
        "identifier",
        transformTest("export-default-declaration/identifier.js"),
      );

      test(
        "value declarations",
        transformTest("export-default-declaration/value-declaration.js"),
      );
    });

    describe("export named declarations", () => {
      test(
        "array pattern",
        transformTest("export-named-declaration/array-pattern.js"),
      );

      test(
        "assignment pattern",
        transformTest("export-named-declaration/assignment-pattern.js"),
      );

      test("class", transformTest("export-named-declaration/class.js"));

      test(
        "identifier",
        transformTest("export-named-declaration/identifier.js"),
      );

      test("function", transformTest("export-named-declaration/function.js"));

      test(
        "object pattern",
        transformTest("export-named-declaration/object-pattern.js"),
      );

      test(
        "rest element",
        transformTest("export-named-declaration/rest-element.js"),
      );
    });

    test(
      "export named from source",
      transformTest("export-named-from-source.js"),
    );

    test(
      "export named specifiers",
      transformTest("export-named-specifiers.js"),
    );

    test("throws when there is an export all declaration", () => {
      expect(() =>
        transform("'use server'; export * from 'pkg';"),
      ).toThrowError();
    });
  };

  function transformTest(filepath: string) {
    return async () => {
      const code = await fs.readFile(
        path.resolve(import.meta.dirname, "fixtures", filepath),
        "utf-8",
      );

      const output = transform(code);

      expect(output).toMatchSnapshot();
    };
  }

  function transform(code: string) {
    const ast = parseAst(code);

    baseTransform(ast, options);

    return generate(ast);
  }
}
