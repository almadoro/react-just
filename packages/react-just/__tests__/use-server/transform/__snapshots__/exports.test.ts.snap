// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms exports on a no-treeshake environment > export default declarations > class declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
class A extends Function {}
const $$Ref$$default = $$registerServerReference$$(A, "action:default", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > function declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
function $$default$$() {}
const $$Ref$$default = $$registerServerReference$$($$default$$, "action:default", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > identifier 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let a;
const $$Ref$$default = $$registerServerReference$$(a, "action:default", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export default declarations > value declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
const $$default$$ = () => {};
const $$Ref$$default = $$registerServerReference$$($$default$$, "action:default", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > array pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let [a, {b}, [c], d = 1, , ...e] = [];
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$(b, "action:b", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerServerReference$$(c, "action:c", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerServerReference$$(d, "action:d", "d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerServerReference$$(e, "action:e", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > assignment pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let [a = 1, {b} = c, [d] = e] = [];
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$(b, "action:b", "b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerServerReference$$(d, "action:d", "d");
export {$$Ref$$d as d};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > class 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
class A extends Function {}
const $$Ref$$A = $$registerServerReference$$(A, "action:A", "A");
export {$$Ref$$A as A};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > function 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
function a() {}
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > identifier 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let a;
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > object pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let {a, b: {c}, d: [e], f = 1, ...g} = {};
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$(c, "action:c", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerServerReference$$(e, "action:e", "e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerServerReference$$(f, "action:f", "f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerServerReference$$(g, "action:g", "g");
export {$$Ref$$g as g};
"
`;

exports[`transforms exports on a no-treeshake environment > export named declarations > rest element 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let [...a] = [];
let [...[b, c]] = [];
let [...{d: e}] = [];
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$(b, "action:b", "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerServerReference$$(c, "action:c", "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerServerReference$$(e, "action:e", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named from source 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
import {a as $$a$$, b as $$c$$, b as $$d$$, default as $$default$$, default as $$e$$} from "pkg";
let c = 1;
const $$Ref$$a = $$registerServerReference$$($$a$$, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$($$c$$, "action:c", "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerServerReference$$($$d$$, "action:d", "d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerServerReference$$($$default$$, "action:default", "default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerServerReference$$($$e$$, "action:e", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a no-treeshake environment > export named specifiers 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/no-treeshake";
let a;
let b;
let d;
const $$Ref$$a = $$registerServerReference$$(a, "action:a", "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$(b, "action:c", "c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerServerReference$$(d, "action:default", "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > class declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > function declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > identifier 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export default declarations > value declarations 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > array pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$("action:b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerServerReference$$("action:c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerServerReference$$("action:d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerServerReference$$("action:e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > assignment pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$("action:b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerServerReference$$("action:d");
export {$$Ref$$d as d};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > class 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$A = $$registerServerReference$$("action:A");
export {$$Ref$$A as A};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > function 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > identifier 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > object pattern 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$("action:c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerServerReference$$("action:e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerServerReference$$("action:f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerServerReference$$("action:g");
export {$$Ref$$g as g};
"
`;

exports[`transforms exports on a treeshake environment > export named declarations > rest element 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerServerReference$$("action:b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerServerReference$$("action:c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerServerReference$$("action:e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named from source 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$("action:c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerServerReference$$("action:d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerServerReference$$("action:e");
export {$$Ref$$e as e};
"
`;

exports[`transforms exports on a treeshake environment > export named specifiers 1`] = `
"import {registerServerReference as $$registerServerReference$$} from "react-just/treeshake";
const $$Ref$$a = $$registerServerReference$$("action:a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerServerReference$$("action:c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerServerReference$$("action:default");
export {$$Ref$$default as default};
"
`;
